### DISCUSSION TITLE
Come integrare uno Shelly EM in HA tramite MQTT

### DISCUSSION THREAD

**Discussion by cpighin** [indomus/forum#2435] [HUB: Home Assistant]
Ciao,

non essendo esperto in domotica e neofita di Home Assistant, ho tentato invano di integrare uno Shelly EM in HA attraverso MQTT. Ho seguito le guide inDomus, ma sono riuscito solo a verificare che il componente MQTT broker funziona (ho fatto prove con i pannelli Pubblica e Ascolta nel percorso Impostazioni > Dispositivi e servizi > Integrazioni > MQTT > Configura ) ma non riesco a leggere le telemetrie dello Shelly.
Sospetto, che lo Shelly EM possa avere qualche problema, dato che non è più raggiungibile col browser al suo indirizzo IP, come accadeva prima che ne modificassi l'impostazione per farlo dialogare in MQTT.
Cosa consigliate di fare, prima che io tenti di resettarlo e riconfigurarlo da capo?

Grazie
Claudio

**Comment by snakuzzo** [comment]
Se il device non è raggiungibile via browser, immagino che non riesca a collegarsi alla tua WIFI.
L'abilitazione del protocollo MQTT non disabilità la connessione WIFI, così come funzionava prima deve funzionare pure dopo.
L'unica cosa che succede abilitando MQTT (con i device di 1a generazione) è la disabilitazione del Cloud di Shelly.
Visto che non lo raggiungi più tramite IP, prova a resettare il device

**Comment by cpighin** [comment]
snacuzzo,

ho resettato lo ShellyEM e, utilizzando la procedura di collegarsi alla sua rete Wifi e ciò che segue, l'ho incluso nella rete Wifi domestica. 
Poi ho aperto una pagina web all'IP dello ShellyEM, ho acceduto alla sezione Internet & Security, poi a Advanced – Developer Settings e lì **ho trovato già abilitata l'opzione Enable MQTT** e già impostati i campi Username, password , Server .... e in “Use Custom MQTT Prefix” il nome dello shelly che avevo impostato nella precedente inclusione. Infine **ho salvato l'impostazione, anche se non avevo modificato nulla, e l'app ha richiesto il riavvio**.

Risultato: ho ricevuto un avviso che **non si era potuto riavviare il dispositivo**, che era diventato Offline e non era più raggiungibile al suo IP. **È sembrato che il tentativo di riavvio dello ShellyEM, pur non modificando le impostazioni MQTT che avevo trovato già impostate, abbia di nuovo reso irraggiungibile il dispositivo al suo IP**.

Come può spigarsi quanto accaduto?
Che ne dici di riprovare l'inclusione, accettare le impostazioni MQTT che presumo rimangano quelle della prima volta, e non riavviare per monitorare cosa succede?

**Reply by snakuzzo** [reply]
Non so che dirti, sinceramente.
Domanda... che nome hai impostato la parte MQTT ? Non è che hai scelto nomi troppo lunghi o con caratteri particolari?

Ad ogni modo, nelle impostazione prova ad abilitare il debug e scaricare i log per capire cosa accade...

![image](https://user-images.githubusercontent.com/7205371/232462861-48b43624-425b-4ecf-8602-79beb571290e.png)

**Reply by snakuzzo** [reply]
@cpighin ho eliminato le tue risposte perché NON devi fare reply direttamente dalla mail e DEVI rispondere direttamente sotto l'ultima risposta evitando di creare nuovi thread (come hai fatto prima).
Grazie.

**Comment by cpighin** [comment]
OK ...come utente MQTT ho scelto mqtt_user...

**Reply by snakuzzo** [reply]
Aridaje, ti avevo chiesto di fare reply sotto l'ultimo messaggio!
Ad ogni, modo...non mi pare ci sia nulla di sbagliato, prova a guardare i log abilitando il debug...ti ho postato lo screenshot...lo trovi sotto settings

**Comment by cpighin** [comment]
Grazie snakuzzo per le risposte precedenti,
      
Ecco le novità. Ho incluso ancora una volta il dispositivo nella rete domestica e, atteso un pò finchè il dispositivo fosse tornato di nuovo raggiungibile al suo IP, ho verificato che erano già presenti le impostazioni MQTT in Advanced – Developer Settings, ma ho notato una stranezza (almeno per me):
nel browser del PC all'IP del device e  nell'app SchellyCloud del cellulare vedevo aggiornarsi i valori del consumo sui 2 canali, mentre tali canali apparivano Offline (in grigio) sul PC all'indirizzo https://home.shelly.cloud/#/....
Sapresti spiegare tale stranezza, prima che io faccia altre azioni?
Nella mia ignoranza, ritengo che lo ShellyEM sia incluso e funzionante (non Offline altrimenti sarebbe riportato in bianco) e forse posso provare a non salvare le impostazioni, in quanto non ho apportato modifiche, per verificare se il dispositivo comunica telemetrie.
In Device Info il Debugging era attivato.

**Comment by cpighin** [comment]
Forse può essere utile sapere che ho provato a scaricare il precedente log e quello corrente e, in entrambi i casi, si è aperta una nuova pagina web, priva di contenuto ...

**Reply by Scialla** [reply]
Cosa non ti è chiaro nel "devi fare reply all'ultimo messaggio per non creare nuovi thread?"
In sostanza devi cliccare su "write a reply" e NON su "write a comment"

Per quanto chiesto prima, quando abiliti mqtt sullo shelly, disabiliti il cloud; è quindi normale che lato cloud non si aggiornino più i valori

**Reply by cpighin** [reply]
Scialla,
ti ringrazio per la precisazione su come proseguire correttamente la discussione.
Nel merito del log, confermi che hai voluto intendere che, essendo disabilitato il Cloud Shelly, Home Assistant non può fornire log, in quanto i dati vengono registrati in tale Cloud e non in HA?

**Reply by Scialla** [reply]
Mi riferivo a quanto dicevi prima, ovvero alla frase: "nel browser del PC all'IP del device e nell'app SchellyCloud del cellulare vedevo aggiornarsi i valori del consumo sui 2 canali, mentre tali canali apparivano Offline (in grigio) sul PC all'indirizzo https://home.shelly.cloud/#/....
Sapresti spiegare tale stranezza, prima che io faccia altre azioni?"
se hai abilitato mqtt in locale, disabilita in automatico la comunicazione col cloud (https://home.shelly.cloud/#/....)

**Reply by cpighin** [reply]
Grazie Scialla.

Dato che non riesco ad andare avanti nella capacità di utilizzare lo ShellyEM in HA tramite MQTT, ho pensato di soprassedere e di provare con un altro dispositivo disponibile, uno Shelly1PM.
Se riuscirò nell'impresa con lo Shelly1PM, spero di avere elementi per capire cosa non va con lo ShellyEM.
Vi sembra una buona idea?

**Reply by snakuzzo** [reply]
Sinceramente, no.
Perché dovresti capirci di più utilizzando un 1PM ?
La logica è esattamente la stessa.
Piuttosto, visto che ormai l'integrazione nativa è stabile, potresti anche decidere di integrarli in questo modo, non trovi ? Così intanto lo integri e lo fai funzionare.
Fatto ciò puoi anche decidere di integrarlo "anche" via MQTT creando delle entità separate...in questo modo smanetti e cerchi di capire avendo comunque tutto integrato

**Reply by cpighin** [reply]
Prima che snacuzzo scrivesse il post precedente, ero alle prese con l'inclusione di uno Shelly1PM non ancora utilizzato, in Shelly Cloud e ... **sorpresa**, ho notato per caso che sull'app HA del cellulare, in Impostazioni > Dispositivi e servizi > Integrazioni erano **"Rilevati "3 shelly da configurare**: 
2 shelly1pm ed 1 shellyem.
Dunque, l'integrazione in HA era vicina! 
Si trattava di uno Shelly1PM la cui integrazione avevo già effettuato mesi fa, del nuovo Shelly1PM e, sorpresa più grande, dello **ShellyEM che da mesi tentavo di integrare**!
Non so spiegare come sia potuto accadere, ma ormai non cerco più spiegazione per fatti strani in informatica ed accolgo con un sorriso quanto accaduto :)
Dopo, senza alcun problema, ho potuto configurare i 2 Shelly1PM ed ora mi accingo a farlo con lo ShellyEM, il quale, a differenza dei PM, è configurata in Shelly Cloud per il servizio MQTT.
Vediamo come va e cosa imparerò ...

**Reply by snakuzzo** [reply]
Secondo me però stai confondendo un po' le cose.
Shelly Cloud e servizio MQTT c'entrano come le alici con la nutella (e te l'abbiamo già spiegato più e più volte).
Quella cosa che ti "ha sorpreso" è una cosa del tutto normale e stiamo parlando dell'integrazione nativa di cui ti parlavo prima.

**Reply by Scialla** [reply]
..che utilizzando l'autodiscovery, ha trovato i tuoi device.. nessuna sorpresa

**Comment by cpighin** [comment]
È "l'integrazione nativa di cui ti parlavo prima", di cui snakuzzo ha scritto, che non era riuscita in passato. Ora, per ragioni che non so, ha funzionato. 
Può benissimo darsi che in passato io abbia sbagliato qualcosa, ma è un fatto positivo che ora abbia funzionato!
Aggiungo che ho configurato con successo in HA anche lo ShellyEM e che ho potuto accedere a vari dati e grafici, cosa di cui sono molto contento.
Da quanto ho potuto capire, ora lo  ShellyEM è regolarmente integrato e sembra in condizione di fornire telemetrie a MQTT, cosa di cui non sento più il bisogno, ma potrei provarci per imparare qualche altra cosa di HA.

**Comment by massimoboscolo** [comment]
Salve, sono un neofita anch'io, una domanda ma per utilizzare uno Shelly 3EM PRO con MQTT devo necessariamente installare ShellyForHass? Con la sua integrazione nativa non funziona?

**Reply by Scialla** [reply]
Tutti gli Shelly sono compatibili nativamente con mqtt, la compatibilità è nel firmware e non nelle integrazioni per homeassistant

**Reply by indomus** [reply]
> Salve, sono un neofita anch'io, una domanda ma per utilizzare uno Shelly 3EM PRO con MQTT devo necessariamente installare ShellyForHass? Con la sua integrazione nativa non funziona?

basta usare l'integrazione "ufficiale" di home assistant

**Comment by massimoboscolo** [comment]
Non capisco come inserire il mio Shelly 3EM Pro nella scheda ENERGY, ho cercato su internet ed ho provato ma non mi esce nulla da selezionare.

**Reply by Scialla** [reply]
Per vederlo nel pannello energia, deve essere tracciato nel recorder, deve avere `device_class: energy` e anche `state_class: total_increasing`

Se non li hanno e il sensore `energy` non è modificabile perché fornito dall'integrazione shelly, puoi aggiungerli nel `customize.yaml`

PS gli attributi li vedi nei dev_tools

**Comment by massimoboscolo** [comment]
@Scialla intendi così?
![1](https://github.com/user-attachments/assets/fd13a2ed-fb5f-4ca5-9b7b-e25e6c58dbfb)

**Reply by Scialla** [reply]
no, devi aggiungere in `customize.yaml` gli attributi che servono al tuo sensore, così (questa è la customizzazione del mio shelly):
```
sensor.casa_energia:
  state_class: total_increasing
  device_class: energy
```

e il sensore deve essere aggiunto nel recorder

**Comment by massimoboscolo** [comment]
@Scialla adesso ho modificato così, ma continuo a non vederlo sulla scheda energy, cosa mi manca?

sensor.shellypro3em_34987a46ec40_em0_power:
  state_class: total_increasing
  device_class: energy
  
sensor.shellypro3em_34987a46ec40_em1_power:
  state_class: total_increasing
  device_class: energy
  
sensor.shellypro3em_34987a46ec40_em2_power:
  state_class: total_increasing
  device_class: energy

**Reply by Scialla** [reply]
Occhio che devi usare il sensore `energy` e non il `power` ovvero quello in Wh e non quello in W
Poi devi metterlo nel recorder e riavviare, lo hai fatto?
Se si devi aspettare un po', deve crearsi un minimo di storico

**Reply by massimoboscolo** [reply]
Ma io non ho il sensore energy, ho solo il power che mi dice i W istantanei, devo crearlo il sensore energy?

**Reply by Scialla** [reply]
Certo! Se non lo hai devi crearlo 
Se poi hai anche fotovoltaico, e quindi lo shelly EM ti va in negativo, devi pure filtrare i valori... comunque, per creare il sensore energy leggi qui:
https://indomus.it/guide/calcolare-il-consumo-di-energia-in-kwh-a-partire-dalla-potenza-istantanea-in-watt-con-home-assistant/

**Reply by massimoboscolo** [reply]
Così non gli piace
![2](https://github.com/user-attachments/assets/15dd6b8f-1650-4a74-a7b2-ccd84a4e01a7)

mi da quest'errore
duplicated mapping key (20:1)

 17 |     unit_prefix: k
 18 |     round: 3
 19 | 
 20 | sensor:
------^
 21 |   - platform: integration
 22 |     source: sensor.shellypro3em ...

**Reply by Scialla** [reply]
si però ci devi arrivare da solo a queste cose.... non puoi mettere `sensor:` tre volte, si mette una volta sola e ci metti poi i tre sensori, altrimenti usa un file sidecar oppure un package

**Reply by massimoboscolo** [reply]
Scusami hai non ragione ma di più, però purtroppo sono entrato nel mondo HA da poco e sto provando, non ho mai programmato nulla ed ovviamente le mie lacune sono voraggine :(
Se non ho capito male sotto sensor devo dichiarare tutti i sensori che voglio creare adesso e anche quelli che creerò in futuro, è corretto?

**Reply by Scialla** [reply]
Si
